### ISSUE TITLE
PBR support

### ISSUE THREAD

**Issue by hbeni** [HHS81/c182s#516]
PBR is the next FGFS renderer and we should make the plane compliant to it some time...

**Comment by TheFGFSEagle** [comment]
HDR, not PBR - maybe you could correct that as it might be confusing to others ! ;)

**Comment by hbeni** [comment]
Sure :)

if you need to touch the model, be aware that there are currently branches modifying that too.

**Comment by hbeni** [comment]
Question: how are we organizing this, should the pbr‘ed planes go into a separate branch or are they downward compatible so we can just merge into master?

**Comment by wlbragg** [comment]
@hbeni im not sure how to best organize a changeover. Here is what I have come across after starting on this with the c172p, j3cub and the Aircrane.

First, I don't know the details about how one is to use and organize PBR textures. Can or are we going to take the same textures we currently use and convert them to PBR textures. How? you have a texture sheet currently that has plastics, metals, rubber, etc., all on the same texture. The Aircrane interior, for example, I think is currently the farthest along of any aircraft so far that has had PBR texturing applied. What I have done so far, right or wrong, is to take a stand alone PBR ORM textures, typically 2048x2048 and uv map some interior object to it, like scratched-shiny-metal. I do the same for say the leather seats using a different PBR black-leather texture. I continued doing this for many of the interior parts, panels, flooring, seat frames. Those PBR textures come complete with the other needed support textures types, shineness, normal, etc.
Now this would continue to be done with every object in that is the aircraft. This would leave you with a ton of textures and now way to organize pint kits and the likes.
So what is another option? Take and existing standard texture and upgrade it to a PBR texture? How? Can that be done? Can you have black-leather, brown-leather, worn-shiny-metal, rusted-metal, rubber, plastic, etc. all on the same texture sheet and then make the shininess and normal map textures from that sheet? Can you leave the existing uv-mappings of multiple objects and apply all those supporting PBR textures using the effects definitions to the same multi-PBR-texture sheet? How much work will that be and who is going to do it?
That is the first questions that came to my mind as I started doing this?

**Comment by wlbragg** [comment]
Thought:

I guess if I were to test the idea of converting an existing texture, I would bring it into gimp with a uv outline map of existing mappings (can be created in Blender) as layer to use for a guide, then replace the existing image with the ORM texture of choice for each different texture mapped on that sheet. Once that is done, then make the shineness and normal textures to match? Is this possible, is this a reasonable way to do this? When done you would still be able to use things like paint kits, what kind of conversions would they need? It's going to require some thought on what is to be the best practice here.

**Comment by CyberGamer224466** [comment]
It'd be way easier to apply separate texture maps for different parts of the model. Then you just have to habe different effect files for each object that uses a different texture but if some are using the same texture they can use the same effect files. I use this approach.

I think that for now it would be way more work if we'd try to fit all different texture maps on one map.

Also the PBR'ed aircraft aren't backwards compatible, at least for now. The ALS shader should be able to also display the same maps as HDR and then we could have real compatibility. Or the second option would be that the aircraft would detect when you're using HDR and when ALS and based on that swap effect files or models that it's using.

**Comment by CyberGamer224466** [comment]
Also, I think that if we're going to go PBR we would need to create a new PBR branch.

**Comment by wlbragg** [comment]
Yes, on the PBR branch. I don't think we will be able to merge it into master as they are going to likely be really different textures as we are discussing.

But so far the question of paint kits and liveries have not been addressed. So I agree somewhat with @CyberGamer about the separate texture maps for the different parts of the model, that is what I have been doing so far for all aircraft I am applying PBR to. But I still don't have a definitive opinion as to whether or not that is the best practice. It is the easiest as far as converting one of these existing aircraft to look good. But it is going to make everything else harder, we haven't even discussed what that does to the creation of lightmaps, if we even are going to have them. I think this all needs to be hashed out before anyone starts in ernest to convert any of these existing models to a production level PBR Rendered model. I did it the way I did it strictly to test it and see immediate results, not as a blueprint for others to follow or with any real though as to what issues it will raise for other disciplines.

**Comment by TheFGFSEagle** [comment]
Yeah, PBR is not backwards compatible at all, at least if you have to modify textures for it - in any case, using model-pbr is not  since that effect is simply not available on non-PBR. 

@CyberGamer224466 since you seem to know a whole lot about this whole PBR stuff - why do any textures even need to modified at all ?

**Comment by wlbragg** [comment]
In regards to a normal map, yes we do, for the exterior for example. If we convert the existing texture, (not the normal map), but the texture we use for the exterior and the ones for the livaries to PBR then yes the normal map we have would work. But if we have to break down the main single texture currently used into different PBR specifice textures such as one for rubber, one for shiny metal, one for plastic, one for fiberglass, etc., then the normal maps have to be made for each on as well. In our case a ORM texture is the base texture (albedo) + a shineness and normal texture combined into a 3 channel RGB texture.

See https://marmoset.co/posts/physically-based-rendering-and-you-can-too/

**Comment by wlbragg** [comment]
For example, how do you convert the rubber in this current default texture to a PBR, ORM style texture.

Current defaul texture for rubber tires.
![defaul](https://github.com/HHS81/c182s/assets/11069680/02d9bafe-f2e5-42cd-bb93-2fa6d383a186)

PBR / ORM textures for rubber.
![synth-rubber-albedo](https://github.com/HHS81/c182s/assets/11069680/99565292-f125-4715-9624-a190b129f9cb)
![synth-rubber-orm](https://github.com/HHS81/c182s/assets/11069680/6b45f96d-e681-41fa-bab6-40763ea0fcb4)
![synth-rubber-normal](https://github.com/HHS81/c182s/assets/11069680/8db3ab86-8173-49fc-afda-f4e2324aa76d)

**Comment by HHS81** [comment]
> It'd be way easier to apply separate texture maps for different parts of the model. Then you just have to have different effect files for each object that uses a different texture but if some are using the same texture they can use the same effect files. I use this approach.
> 
> I think that for now it would be way more work if we'd try to fit all different texture maps on one map.

Would make things easier, but would mabye needs much more RAM, if you have many different materials. You way was not recommended in the past, and creating texture maps isn't difficult, as Blender has the tools for it.
Btw.: X-Plane uses HDR, pretty much the same way we do now (just much better- and much better framerates!). Can't tell for the Payware aicraft, but the default aircraft uses texture maps, organized in objects and not in materials.
As an example: X-Planes A330
![A330_text_org](https://github.com/HHS81/c182s/assets/12649375/79ebdb17-1135-4166-a511-a611608ba6b8)

> 
> Or the second option would be that the aircraft would detect when you're using HDR and when ALS and based on that swap effect files or models that it's using.

I'm not sure if this possible, if so then maybe in the .eff-files.

**Comment by HHS81** [comment]
> > I'm not sure if this possible, if so then maybe in the .eff-files.
> 
> I could potentially see this being possible starting at the variant definition.
> 
> Variant ALS calls Model/c182s-als.xml Variant PBS calls Model/c182s-pbr.xml
> 
> The differences in these two models are the .ac, textures and the effects definitions applied to them. I'm pretty sure it can be done this way. But do we want to? In the future, we may not even care about anything but the PBR/ORM version. If we get to the point where we only care about the PBR version, then the other variant and its associated files are deleted. It wouldn't be that difficult if we keep the folders and files for each isolate from each other.

Ideed, the better method! 👍

**Comment by CyberGamer224466** [comment]
> For example, how do you convert the rubber in this current default texture to a PBR, ORM style texture.
> 
> Current defaul texture for rubber tires.
> ![defaul](https://github.com/HHS81/c182s/assets/11069680/02d9bafe-f2e5-42cd-bb93-2fa6d383a186)
> 
> PBR / ORM textures for rubber.
> albedo texture
> ![synth-rubber-albedo](https://github.com/HHS81/c182s/assets/11069680/99565292-f125-4715-9624-a190b129f9cb)
> ORM texture
> ![synth-rubber-orm](https://github.com/HHS81/c182s/assets/11069680/6b45f96d-e681-41fa-bab6-40763ea0fcb4)
> normal texture
> ![synth-rubber-normal](https://github.com/HHS81/c182s/assets/11069680/8db3ab86-8173-49fc-afda-f4e2324aa76d)
> 
> 

Yeah, it'd be quite difficult to convert thus texture to separate smaller ones as we need a lot of different maps like your examples show, ORM, normal etc. Also the resolution is a big issue cause if we'd do it this way it'd be way too small.

**Comment by wlbragg** [comment]
>Also the resolution is a big issue cause if we'd do it this way it'd be way too small.

Right, but this appears to be what x-plane is doing. I don't know how large their textures are, but a 4096 would go a long way to alleviate the resolution issue. Also a possible remapping of some object may be needed so you can get higher resolutions on objects that need it.

**Comment by hbeni** [comment]
I have read about the gltf format which a recent blender should be able to produce out of the box.
If I have read that correctly, that export would produce all that is needed, and the materials are defined in the blender file. That I think does not need new textures.
So if i understand correctly, that may be a less work intensive way to convert to HDR/PBR.

But I'm really out of this 3D stuff, and as I have a ATI card I cant assist in testing (yet).
Also I would suggest we don't do the variant way:
- In my view its easier to just maintain the "master" branch here and have the most compatible code here. For people using PBR, they can easily download the pbr'ed plane from here.
- Additionally we should have a "PBR-master" branch where we do the advanced stuff. New features to "master" can easily be merged into that branch.
- Finally, when HDR/PBR gets default in flightgear (which I strongly assume it will some day), we will be easily able to merge the PBR branch into master and drop it, at which point we just need to maintain one single branch again.

**Comment by TheFGFSEagle** [comment]
That sounds like a good plan; also I can assist with HDR testing since I got my external Nvidia Quadro K620 to work (need to find a way to correct the overexposure first, though)

**Comment by hbeni** [comment]
For that matter - I still look for someone assisting with the HDR/PBR conversion....

---- 
As FGFS release plans clarified that HDR/PBR is the only supported rendering for `next`, I think we should go this way:
* I will split off a 2024.1 Branch (currently planned stable for end of this year), which we can maintain the non-PBR version.
* The `master` branch will still track the FGFS `next`, like it did for quite some years now.

I will face the issue that I will get stuck to 2024.1 for the time beeing (as my ATI card is not yet supported by FGFS). So I will branch off my developments from there and try to maintain the "official release" and keep also FGAddon releases current. Changes I will merge "up" to `next`, so we have them also in the mainline.

What are your toughts about that?

**Comment by hbeni** [comment]
I hope they make progression so I can fly with hdr/pbr with my ati card 😉

**Comment by stuartbuchanan** [comment]
Hi Folks,


Just to say that IMO that is an excellent plan and aligned with how we are
expecting aircraft development to progress.

Gltf is the way forward for PBR/HDR

Stuart

**Comment by wlbragg** [comment]
A bit more info here. I decided to try to import the simi hdr/pbr version of the c172p I have worked on as a test case into Blender 3.6.0 and export it as a .gtlf. Using 2020.4 (I don't have a working build of 2024 next yet). I don't see how using the gtlf features of Blender are really going to make much difference in the questions asked here. From what I can tell, you are still working with pbr texturing as describe in this thread. Blender internal features are better or more robust using this format, but material handling, shader and mesh work are all still required, in fact more so in the gltf format. The workflow and general model and texture creation (as in having to create textures in some format) however are basically the same. So the questions on "best practices" being asked here still apply.

I was successful in importing the HDR/PBR version I have into 3.6.0 and then exporting it to Gltf. The only issue was the transparent mesh came out not transparent in the sim, also the hidden (transparent) hot model mesh we use to expand a pick animation hot area. That may be just a matter of Blender Gltf settings. I need to spend time and energy to research all that might be involved.

The bottom line is that
>Gltf is the way forward for PBR/HDR

is accurate, but, it should not be taken to imply that the questions we are asking here will vanish because of this format. In fact it will add to this discussion.
